<a href="https://colab.research.google.com/github/tanerijun/ml-training/blob/main/numpy_pandas_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://www.hackerearth.com/practice/machine-learning/data-manipulation-visualisation-r-python/tutorial-data-manipulation-numpy-pandas-python/tutorial/

In [162]:
import numpy as np
import pandas as pd

In [163]:
np.__version__

'1.25.2'

In [164]:
# Create a list of numbers from 0 to 9
L = list(range(10))

In [165]:
# Converting integers to string
[str(c) for c in L]

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']

In [166]:
[type(item) for item in L]

[int, int, int, int, int, int, int, int, int, int]

# Creating Arrays

Numpy arrays are homogeneous in nature, i.e., they comprise of one data type (int, float, double, etc.) unlike lists.

In [167]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [168]:
np.zeros(10, dtype='int')

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [169]:
# creating a 3 x 5 matrix
np.ones((3, 5), dtype=float)

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [170]:
# creating a matrix with predefined value
np.full((3, 5), 1.23)

array([[1.23, 1.23, 1.23, 1.23, 1.23],
       [1.23, 1.23, 1.23, 1.23, 1.23],
       [1.23, 1.23, 1.23, 1.23, 1.23]])

In [171]:
# create an array with a set sequence
a = np.arange(0, 20)
b = np.arange(20)
c = np.arange(0, 20, 2)
print(a)
print(b)
print(c)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]
[ 0  2  4  6  8 10 12 14 16 18]


In [172]:
# create an array of even space between the given range of values
np.linspace(0, 1, 5) # linear space

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

In [173]:
# create a 3x3 array with mean 0 and standard deviation 1 in a given dimension
np.random.normal(0, 1, (3, 3))

array([[-0.2509793 , -0.41522808, -0.63908615],
       [ 2.49776759, -3.00512816, -0.87399723],
       [-2.65018924,  1.12400545, -0.21363701]])

In [174]:
# create a 3x3 array from uniform distribution
a = np.random.rand(3, 3) # range 0 to 1
b = np.random.uniform(0, 2, (3, 3)) # allows specifying the interval
print(a)
print(b)

[[0.82687037 0.50934207 0.08637698]
 [0.66003986 0.20659461 0.84727532]
 [0.68135913 0.17836735 0.06985902]]
[[0.01937634 1.78409932 0.26693042]
 [1.55838395 1.85032685 1.4303575 ]
 [0.98163723 0.93899661 1.76541858]]


In [175]:
# create an identity matrix
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [176]:
# flip matrix
np.fliplr(np.eye(3))

array([[0., 0., 1.],
       [0., 1., 0.],
       [1., 0., 0.]])

In [177]:
# set a random seed
np.random.seed(0)

In [178]:
x1 = np.random.randint(10, size=6) # 1D
x2 = np.random.randint(10, size=(3, 4)) # 2D
x3 = np.random.randint(10, size=(3, 4, 5)) # 3D

print(x1)
print(x2)
print(x3)

[5 0 3 3 7 9]
[[3 5 2 4]
 [7 6 8 8]
 [1 6 7 7]]
[[[8 1 5 9 8]
  [9 4 3 0 3]
  [5 0 2 3 8]
  [1 3 3 3 7]]

 [[0 1 9 9 0]
  [4 7 3 2 7]
  [2 0 0 4 5]
  [5 6 8 4 1]]

 [[4 9 8 1 1]
  [7 9 9 3 6]
  [7 2 0 3 5]
  [9 4 4 6 4]]]


In [179]:
print("x3 ndim:", x3.ndim)
print("x3 shape:", x3.shape)
print("x3 size:", x3.size)

x3 ndim: 3
x3 shape: (3, 4, 5)
x3 size: 60


## Array Indexing

In [180]:
x1 = np.array([4, 3, 4, 4, 8, 4])
x1

array([4, 3, 4, 4, 8, 4])

In [181]:
print(x1[0])
print(x1[4])
print(x1[-1])
print(x1[-2])

4
8
4
8


In [182]:
x2

array([[3, 5, 2, 4],
       [7, 6, 8, 8],
       [1, 6, 7, 7]])

In [183]:
print(x2[2, 3])
print(x2[2, -1])
print(x2[0, 0])

7
7
3


## Array Slicing

In [184]:
# x = np.arange(10)
# x
# array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])


#from start to 4th position
# x[:5]
# array([0, 1, 2, 3, 4])


#from 4th position to end
# x[4:]
# array([4, 5, 6, 7, 8, 9])


#from 4th to 6th position
# x[4:7]
# array([4, 5, 6])


#return elements at even place
# x[ : : 2]
# array([0, 2, 4, 6, 8])


#return elements from first position step by two
# x[1::2]
# array([1, 3, 5, 7, 9])


#reverse the array
# x[::-1]
# array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

## Array Concatenation
Many a time, we are required to combine different arrays. So, instead of typing each of their elements manually, you can use array concatenation to handle such tasks easily.

In [185]:
# you can concatenate 2 or more arrays at once
x = np.array([1, 2, 3])
y = np.array([3, 2, 1])
z = [21, 21, 21]
np.concatenate([x, y])

array([1, 2, 3, 3, 2, 1])

In [186]:
np.concatenate([x, y, z])

array([ 1,  2,  3,  3,  2,  1, 21, 21, 21])

In [187]:
# You can also use this function to create 2D arrays
grid = np.array([[1, 2, 3], [4, 5, 6]])
print(grid)
print(np.concatenate([grid, grid]))

[[1 2 3]
 [4 5 6]]
[[1 2 3]
 [4 5 6]
 [1 2 3]
 [4 5 6]]


In [188]:
# Using its axis parameter, you can define row-wise or column-wise matrix
np.concatenate([grid, grid], axis=1)

array([[1, 2, 3, 1, 2, 3],
       [4, 5, 6, 4, 5, 6]])

Until now, we used the concatenation function of arrays of equal dimension. But, what if you are required to combine a 2D array with 1D array? In such situations, np.concatenate might not be the best option to use. Instead, you can use np.vstack or np.hstack to do the task. Let's see how!

In [189]:
x = np.array([3, 4, 5])
grid = np.array([[1, 2, 3], [17, 18, 19]])
print(np.vstack([x, grid]))
print(np.vstack([grid, x]))

[[ 3  4  5]
 [ 1  2  3]
 [17 18 19]]
[[ 1  2  3]
 [17 18 19]
 [ 3  4  5]]


In [190]:
# Similarly, you can add an array using np.hstack
z = np.array([[9], [9]])
np.hstack([grid, z])

array([[ 1,  2,  3,  9],
       [17, 18, 19,  9]])

We can also split the arrays based on predefined positions

In [191]:
x = np.arange(10)
x

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [192]:
x1, x2, x3 = np.split(x, [3, 6])
print(x1, x2, x3)

[0 1 2] [3 4 5] [6 7 8 9]


In [193]:
grid = np.arange(16).reshape((4, 4))
grid

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [194]:
upper, lower = np.vsplit(grid, [2])
print(upper)
print(lower)

[[0 1 2 3]
 [4 5 6 7]]
[[ 8  9 10 11]
 [12 13 14 15]]


## Arithmetic

In [195]:
a = np.array([10, 20, 30])
b = np.array([2, 5, 3])
result = np.divide(a, b)
print(result)

[ 5.  4. 10.]


In [196]:
result = np.multiply(a, b)
result

array([ 20, 100,  90])

In [197]:
print(a / 5)
print(b * 3)

[2. 4. 6.]
[ 6 15  9]


In [198]:
a = np.array([-1, -2, 3])
result = np.abs(a)
result

array([1, 2, 3])

In [199]:
a = np.array([2, 3, 4])
result = np.power(a, 2)
result

array([ 4,  9, 16])

In [200]:
result = np.mod(a, 2)
result

array([0, 1, 0])

In [201]:
a = np.array([0, np.pi/2, np.pi])
result = np.sin(a)
result

array([0.0000000e+00, 1.0000000e+00, 1.2246468e-16])

In [202]:
result = np.cos(a)
result

array([ 1.000000e+00,  6.123234e-17, -1.000000e+00])

In [203]:
result = np.tan(a)
result

array([ 0.00000000e+00,  1.63312394e+16, -1.22464680e-16])

In [204]:
a = np.array([1, np.e, np.e**2])
result = np.log(a)
result

array([0., 1., 2.])

In [205]:
a = np.array([1, 2, 3, 4, 5])
result = np.var(a) # variance
result

2.0

About variance:

In practice, variance gives you an idea of how spread out the data points are. A high variance indicates that the data points are spread out over a wider range of values, while a low variance indicates that they are closer to the mean.

In [206]:
print(np.min(a))
print(np.max(a))
print(np.mean(a))

1
5
3.0


In [207]:
a = np.array([1, 2, 3])
res = np.exp(a) # e^x
res

array([ 2.71828183,  7.3890561 , 20.08553692])

In [208]:
res = np.sqrt(a)
res

array([1.        , 1.41421356, 1.73205081])

# Pandas

In [209]:
#create a data frame - dictionary is used here where keys get converted to column names and values to row values.
data = pd.DataFrame({
    'Country': ['Russia', 'Colombia', 'Chile', 'Equador', 'Nigeria'],
    'Rank': [121, 40, 100, 130, 11]
})
data

Country  Rank
0    Russia   121
1  Colombia    40
2     Chile   100
3   Equador   130
4   Nigeria    11

In [210]:
# We can do a quick analysis of any data set using:
data.describe()

Rank
count    5.000000
mean    80.400000
std     52.300096
min     11.000000
25%     40.000000
50%    100.000000
75%    121.000000
max    130.000000

describe() method computes summary statistics of integer / double variables. To get the complete information about the data set, we can use info() function.

In [211]:
# Among other things, it shows the data set has 5 rows and 2 columns
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Country  5 non-null      object
 1   Rank     5 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


In [212]:
# Let's create another data frame
data = pd.DataFrame({
    'group': ['a', 'a', 'a', 'b', 'b', 'b', 'b', 'c', 'c', 'c'],
    'ounces': [4, 3, 12, 6, 7.5, 8, 3, 5, 6, 7]
})
data

group  ounces
0     a     4.0
1     a     3.0
2     a    12.0
3     b     6.0
4     b     7.5
5     b     8.0
6     b     3.0
7     c     5.0
8     c     6.0
9     c     7.0

In [213]:
# Let's sort the data frame by ounces - inplace = True will make changes to the data
data.sort_values(by=['ounces'], ascending=True, inplace=False)

group  ounces
1     a     3.0
6     b     3.0
0     a     4.0
7     c     5.0
3     b     6.0
8     c     6.0
9     c     7.0
4     b     7.5
5     b     8.0
2     a    12.0

In [214]:
# We can sort the data by not just one column but multiple columns as well
data.sort_values(by=['group', 'ounces'], ascending=[True, False], inplace=False)

group  ounces
2     a    12.0
0     a     4.0
1     a     3.0
5     b     8.0
4     b     7.5
3     b     6.0
6     b     3.0
9     c     7.0
8     c     6.0
7     c     5.0

Often, we get dataset with duplicate rows, which is nothing but noise. We can handle that by:

In [215]:
# Create a dataframe with duplicate rows
data = pd.DataFrame({
    'k1': ['one'] * 3 + ['two'] * 4,
    'k2': [1, 1, 2, 3, 3, 4, 4]
})
data

k1  k2
0  one   1
1  one   1
2  one   2
3  two   3
4  two   3
5  two   4
6  two   4

In [216]:
data.sort_values(by='k2')

k1  k2
0  one   1
1  one   1
2  one   2
3  two   3
4  two   3
5  two   4
6  two   4

In [217]:
# remove duplicates - ta da!
data.drop_duplicates()

k1  k2
0  one   1
2  one   2
3  two   3
5  two   4

Here, we removed duplicates based on matching row values across all columns. Alternatively, we can also remove duplicates based on a particular column. Let's remove duplicate values from the k1 column.

In [218]:
data.drop_duplicates(subset='k1')

k1  k2
0  one   1
3  two   3

Now, we will learn to categorize rows based on a predefined criteria. It happens a lot while data processing where you need to categorize a variable. For example, say we have got a column with country names and we want to create a new variable 'continent' based on these country names. In such situations, we will require the steps below:

In [219]:
data = pd.DataFrame({
    'food': ['bacon', 'pulled pork', 'bacon', 'Pastrami', 'corned beef', 'Bacon', 'pastrami', 'honey ham', 'nova lox'],
    'ounces': [4, 3, 12, 6, 7.5, 8, 3, 5, 6]
})

In [220]:
data

food  ounces
0        bacon     4.0
1  pulled pork     3.0
2        bacon    12.0
3     Pastrami     6.0
4  corned beef     7.5
5        Bacon     8.0
6     pastrami     3.0
7    honey ham     5.0
8     nova lox     6.0

Now, we want to create a new variable which indicates the type of animal which acts as the source of the food. To do that, first we'll create a dictionary to map the food to the animals. Then, we'll use map function to map the dictionary's values to the keys. Let's see how is it done.

In [221]:
meat_to_animal = {
    'bacon': 'pig',
    'pulled pork': 'pig',
    'pastrami': 'cow',
    'corned beef': 'cow',
    'honey ham': 'pig',
    'nova lox': 'salmon'
}

def meat_2_animal(series):
    if series['food'] == 'bacon':
        return 'pig'
    elif series['food'] == 'pulled pork':
        return 'pig'
    elif series['food'] == 'pastrami':
        return 'cow'
    elif series['food'] == 'corned beef':
        return 'cow'
    elif series['food'] == 'honey ham':
        return 'pig'
    else:
        return 'salmon'

# create a new variable
data['animal'] = data['food'].map(str.lower).map(meat_to_animal)
data

food  ounces  animal
0        bacon     4.0     pig
1  pulled pork     3.0     pig
2        bacon    12.0     pig
3     Pastrami     6.0     cow
4  corned beef     7.5     cow
5        Bacon     8.0     pig
6     pastrami     3.0     cow
7    honey ham     5.0     pig
8     nova lox     6.0  salmon

In [222]:
lower = lambda x: x.lower()
data['food'] = data['food'].apply(lower)
data['animal2'] = data.apply(meat_2_animal, axis=1)
data

food  ounces  animal animal2
0        bacon     4.0     pig     pig
1  pulled pork     3.0     pig     pig
2        bacon    12.0     pig     pig
3     pastrami     6.0     cow     cow
4  corned beef     7.5     cow     cow
5        bacon     8.0     pig     pig
6     pastrami     3.0     cow     cow
7    honey ham     5.0     pig     pig
8     nova lox     6.0  salmon  salmon

In [223]:
# Another way to create a new variable is by using the assign function
data.assign(ounces2 = data['ounces'] * 10)

food  ounces  animal animal2  ounces2
0        bacon     4.0     pig     pig     40.0
1  pulled pork     3.0     pig     pig     30.0
2        bacon    12.0     pig     pig    120.0
3     pastrami     6.0     cow     cow     60.0
4  corned beef     7.5     cow     cow     75.0
5        bacon     8.0     pig     pig     80.0
6     pastrami     3.0     cow     cow     30.0
7    honey ham     5.0     pig     pig     50.0
8     nova lox     6.0  salmon  salmon     60.0

In [224]:
# Remove column animal2
data.drop('animal2', axis=1, inplace=True)
data

food  ounces  animal
0        bacon     4.0     pig
1  pulled pork     3.0     pig
2        bacon    12.0     pig
3     pastrami     6.0     cow
4  corned beef     7.5     cow
5        bacon     8.0     pig
6     pastrami     3.0     cow
7    honey ham     5.0     pig
8     nova lox     6.0  salmon

We frequently find missing values in our data set. A quick method for imputing missing values is by filling the missing value with any random number. Not just missing values, you may find lots of outliers in your data set, which might require replacing. Let's see how can we replace values.

In [225]:
# Series function from pandas are used to create arrays
data = pd.Series([1., -999., 2., -999., -1000., 3.])
data

0       1.0
1    -999.0
2       2.0
3    -999.0
4   -1000.0
5       3.0
dtype: float64

In [226]:
# replace -999 with NaN
data.replace(-999, np.nan, inplace=True)
data

0       1.0
1       NaN
2       2.0
3       NaN
4   -1000.0
5       3.0
dtype: float64

In [227]:
# replace multiple values at once
data = pd.Series([1., -999., 2., -999., -1000., 3.])
data.replace([-999, -1000], np.nan, inplace=True)
data

0    1.0
1    NaN
2    2.0
3    NaN
4    NaN
5    3.0
dtype: float64

In [228]:
# Let's learn how to rename column names and axis (row names)
data = pd.DataFrame(np.arange(12).reshape((3, 4)), index=['Ohio', 'Colorado', 'New York'], columns=['one', 'two', 'three', 'four'])
data

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7
New York    8    9     10    11

In [229]:
# using rename function
data.rename(index = {'Ohio': 'SanF'}, columns={'one': 'one_p', 'two': 'two_p'}, inplace=True)
data

one_p  two_p  three  four
SanF          0      1      2     3
Colorado      4      5      6     7
New York      8      9     10    11

In [230]:
# using string functions
data.rename(index = str.upper, columns = str.title, inplace = True)
data

One_P  Two_P  Three  Four
SANF          0      1      2     3
COLORADO      4      5      6     7
NEW YORK      8      9     10    11

Next, we'll learn to categorize (bin) continuous variables.

In [231]:
ages = [20, 22, 25, 27, 21, 23, 37, 31, 61, 45, 41, 32]

We'll divide the ages into bins such as 18-25, 26-35,36-60 and 60 and above.

In [232]:
# Understand the output - '(' means the value is included in the bin, '[' means the value is excluded
bins = [18, 25, 35, 60, 100]
cats = pd.cut(ages, bins)
cats

[(18, 25], (18, 25], (18, 25], (25, 35], (18, 25], ..., (25, 35], (60, 100], (35, 60], (35, 60], (25, 35]]
Length: 12
Categories (4, interval[int64, right]): [(18, 25] < (25, 35] < (35, 60] < (60, 100]]

In [233]:
# to include the right bin value
pd.cut(ages, bins, right=False)

[[18, 25), [18, 25), [25, 35), [25, 35), [18, 25), ..., [25, 35), [60, 100), [35, 60), [35, 60), [25, 35)]
Length: 12
Categories (4, interval[int64, left]): [[18, 25) < [25, 35) < [35, 60) < [60, 100)]

In [234]:
# check how many observations fall under each bin
pd.value_counts(cats)

(18, 25]     5
(25, 35]     3
(35, 60]     3
(60, 100]    1
Name: count, dtype: int64

In [235]:
# we can pass a unique name to each label
bin_names = ['Youth', 'YoungAdult', 'MiddleAged', 'Senior']
new_cats = pd.cut(ages, bins, labels=bin_names)
print(new_cats)
pd.value_counts(new_cats)

['Youth', 'Youth', 'Youth', 'YoungAdult', 'Youth', ..., 'YoungAdult', 'Senior', 'MiddleAged', 'MiddleAged', 'YoungAdult']
Length: 12
Categories (4, object): ['Youth' < 'YoungAdult' < 'MiddleAged' < 'Senior']


Youth         5
YoungAdult    3
MiddleAged    3
Senior        1
Name: count, dtype: int64

In [236]:
# we can also calculate their cumulative sum
pd.value_counts(new_cats).cumsum()

Youth          5
YoungAdult     8
MiddleAged    11
Senior        12
Name: count, dtype: int64

Let's proceed and learn about grouping data and creating pivots in pandas. It's an immensely important data analysis method which you'd probably have to use on every data set you work with.

In [237]:
df = pd.DataFrame({
    'key1': ['a', 'a', 'b', 'b', 'a'],
    'key2': ['one', 'two', 'one', 'two', 'one'],
    'data1': np.random.randn(5),
    'data2': np.random.randn(5)
})
df

key1 key2     data1     data2
0    a  one  1.254414  1.149076
1    a  two  1.419102 -1.193578
2    b  one -0.743856  1.141042
3    b  two -2.517437  1.509445
4    a  one -1.507096  1.067775

In [238]:
# calc the mean of data1 by key1
grouped = df['data1'].groupby(df['key1'])
grouped.mean()

key1
a    0.388807
b   -1.630647
Name: data1, dtype: float64

In [239]:
# the code above: mean of every data with key1=a
(df['data1'][0] + df['data1'][1] + df['data1'][4]) / 3

0.3888066951970919

In [240]:
# slicing data frame
dates = pd.date_range('20130101', periods=6)
print(dates)
df = pd.DataFrame(np.arange(24).reshape((6, 4)), index=dates, columns=['A', 'B', 'C', 'D'])
df

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06'],
              dtype='datetime64[ns]', freq='D')


A   B   C   D
2013-01-01   0   1   2   3
2013-01-02   4   5   6   7
2013-01-03   8   9  10  11
2013-01-04  12  13  14  15
2013-01-05  16  17  18  19
2013-01-06  20  21  22  23

In [241]:
# get first n rows from df
df[:3]

A  B   C   D
2013-01-01  0  1   2   3
2013-01-02  4  5   6   7
2013-01-03  8  9  10  11

In [242]:
# slice based on date range
df['20130101':'20130104']

A   B   C   D
2013-01-01   0   1   2   3
2013-01-02   4   5   6   7
2013-01-03   8   9  10  11
2013-01-04  12  13  14  15

In [243]:
# slice based on column names
df.loc[:, 'A':'B']

A   B
2013-01-01   0   1
2013-01-02   4   5
2013-01-03   8   9
2013-01-04  12  13
2013-01-05  16  17
2013-01-06  20  21

In [244]:
# slicing based on row and column
df.loc['20130102':'20130104', ['A', 'B']]

A   B
2013-01-02   4   5
2013-01-03   8   9
2013-01-04  12  13

In [245]:
# slicing based on index of columns
df.iloc[3] # returns 4th row

A    12
B    13
C    14
D    15
Name: 2013-01-04 00:00:00, dtype: int64

In [246]:
# returns a specific range of rows
df.iloc[2:4, 0:2]

A   B
2013-01-03   8   9
2013-01-04  12  13

In [247]:
# returns specific rows and columns (not range like the example above)
df.iloc[[2,4], [0,2]]

A   C
2013-01-03   8  10
2013-01-05  16  18

In [248]:
# boolean indexing
df[df.A > 12]

A   B   C   D
2013-01-05  16  17  18  19
2013-01-06  20  21  22  23

In [249]:
# we can copy the dataset
df2 = df.copy()
df2['E'] = ['one', 'one', 'two', 'three', 'four', 'three']
df2

A   B   C   D      E
2013-01-01   0   1   2   3    one
2013-01-02   4   5   6   7    one
2013-01-03   8   9  10  11    two
2013-01-04  12  13  14  15  three
2013-01-05  16  17  18  19   four
2013-01-06  20  21  22  23  three

In [250]:
df # remains unchanged

A   B   C   D
2013-01-01   0   1   2   3
2013-01-02   4   5   6   7
2013-01-03   8   9  10  11
2013-01-04  12  13  14  15
2013-01-05  16  17  18  19
2013-01-06  20  21  22  23

In [251]:
# select rows based on column values
df2[df2['E'].isin(['two', 'four'])]

A   B   C   D     E
2013-01-03   8   9  10  11   two
2013-01-05  16  17  18  19  four

In [252]:
# select all rows except those with two and four
df2[~df2['E'].isin(['two', 'four'])]

A   B   C   D      E
2013-01-01   0   1   2   3    one
2013-01-02   4   5   6   7    one
2013-01-04  12  13  14  15  three
2013-01-06  20  21  22  23  three

We can also use a query method to select columns based on a criterion.

In [253]:
# list all columns where A is greater than C
df.query('A < C')

A   B   C   D
2013-01-01   0   1   2   3
2013-01-02   4   5   6   7
2013-01-03   8   9  10  11
2013-01-04  12  13  14  15
2013-01-05  16  17  18  19
2013-01-06  20  21  22  23

In [254]:
# using OR condition
df.query('A < B | C > A')

A   B   C   D
2013-01-01   0   1   2   3
2013-01-02   4   5   6   7
2013-01-03   8   9  10  11
2013-01-04  12  13  14  15
2013-01-05  16  17  18  19
2013-01-06  20  21  22  23

Pivot tables are extremely useful in analyzing data using a customized tabular format. I think, among other things, Excel is popular because of the pivot table option. It offers a super-quick way to analyze data.

In [255]:
data = pd.DataFrame({
    'group': ['a', 'a', 'a', 'b', 'b', 'b', 'c', 'c', 'c'],
    'ounces': [4, 3, 12, 6, 7.5, 8, 3, 5, 6],
})
data

group  ounces
0     a     4.0
1     a     3.0
2     a    12.0
3     b     6.0
4     b     7.5
5     b     8.0
6     c     3.0
7     c     5.0
8     c     6.0

In [256]:
# calc means of each group
data.pivot_table(values='ounces', index='group', aggfunc=np.mean)

ounces
group          
a      6.333333
b      7.166667
c      4.666667

In [257]:
# calc count by each group
data.pivot_table(values='ounces', index='group', aggfunc='count')

ounces
group        
a           3
b           3
c           3

## Exploring ML Data Set

We'll work with the popular adult data set.The data set has been taken from UCI Machine Learning Repository. You can download the data from [here](https://s3-ap-southeast-1.amazonaws.com/he-public-data/datafiles19cdaf8.zip). In this data set, the dependent variable is "target." It is a binary classification problem. We need to predict if the salary of a given person is less than or more than 50K.

In [258]:
# load data
train = pd.read_csv("/content/train.csv")
test = pd.read_csv("/content/test.csv")

In [259]:
# check dataset
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       30725 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      30718 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  31978 non-null  object
 14  target          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


We see that, the train data has 32561 rows and 15 columns. Out of these 15 columns, 6 have integers classes and the rest have object (or character) classes. Similarly, we can check for test data. An alternative way of quickly checking rows and columns is

In [260]:
print("The train data has", train.shape)
print("The test data has", test.shape)

The train data has (32561, 15)
The test data has (16281, 15)


In [261]:
# let's take a glimpse of the dataset
train.head()

age          workclass  fnlwgt   education  education.num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital.status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital.gain  capital.loss  hours.per.week  native.country  target  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [262]:
# let's check if there's any missing values in this data
nans = train.shape[0] - train.dropna().shape[0]
print("There are", nans, "missing values in the train data.")

There are 2399 missing values in the train data.


In [263]:
nand = test.shape[0] - test.dropna().shape[0]
print("There are", nand, "missing values in the test data.")

There are 1221 missing values in the test data.


In [264]:
# which columns have missing values?
train.isnull().sum() # shows that 3 columns have missing values

age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
target               0
dtype: int64

In [265]:
# Let's count the number of unique values from character variables
cat = train.select_dtypes(include=['O'])
cat.apply(pd.Series.nunique)

workclass          8
education         16
marital.status     7
occupation        14
relationship       6
race               5
sex                2
native.country    41
target             2
dtype: int64

Since missing values are found in all 3 character variables, let's imput these missing values with their respective modes.

In [266]:
# Workclass
train.workclass.value_counts(sort=True)
train.workclass.fillna('Private', inplace=True)

# Occupation
train.occupation.value_counts(sort=True)
train.occupation.fillna('Prof-specialty', inplace=True)

# Native Country
train['native.country'].value_counts(sort=True)
train['native.country'].fillna('United-States', inplace=True)

In [267]:
# let's check if there are any missing values left
train.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
target            0
dtype: int64

Now, we'll check the target variable to investigate if this data is imbalanced or not

In [268]:
# Check proportion of target variable
train.target.value_counts()/train.shape[0]

target
 <=50K    0.75919
 >50K     0.24081
Name: count, dtype: float64

We see that 75% of the data set belongs to <=50K class. This means that even if we take a rough guess of target prediction as <=50K, we'll get 75% accuracy. Isn't that amazing? Let's create a cross tab of the target variable with education. With this, we'll try to understand the influence of education on the target variable.

In [269]:
pd.crosstab(train.education, train.target, margins=True)/train.shape[0]

target            <=50K      >50K       All
education                                  
 10th          0.026750  0.001904  0.028654
 11th          0.034243  0.001843  0.036086
 12th          0.012285  0.001013  0.013298
 1st-4th       0.004975  0.000184  0.005160
 5th-6th       0.009736  0.000491  0.010227
 7th-8th       0.018611  0.001228  0.019840
 9th           0.014957  0.000829  0.015786
 Assoc-acdm    0.024631  0.008139  0.032769
 Assoc-voc     0.031357  0.011087  0.042443
 Bachelors     0.096250  0.068210  0.164461
 Doctorate     0.003286  0.009398  0.012684
 HS-grad       0.271060  0.051442  0.322502
 Masters       0.023464  0.029452  0.052916
 Preschool     0.001566  0.000000  0.001566
 Prof-school   0.004699  0.012991  0.017690
 Some-college  0.181321  0.042597  0.223918
All            0.759190  0.240810  1.000000

We see that out of 75% people with <=50K salary, 27% people are high school graduates, which is correct as people with lower levels of education are expected to earn less. On the other hand, out of 25% people with >=50K salary, 6% are bachelors and 5% are high-school grads. Now, this pattern seems to be a matter of concern. That's why we'll have to consider more variables before coming to a conclusion.

If you've come this far, you might be curious to get a taste of building your first machine learning model. In the coming week we'll share an exclusive tutorial on machine learning in python. However, let's get a taste of it here.

We'll use the famous and formidable scikit learn library. Scikit learn accepts data in numeric format. Now, we'll have to convert the character variable into numeric. We'll use the labelencoder function.

In label encoding, each unique value of a variable gets assigned a number, i.e., let's say a variable color has four values ['red','green','blue','pink'].

Label encoding this variable will return output as: red = 2 green = 0 blue = 1 pink = 3

In [270]:
# load sklearn and encode all object type variables
from sklearn import preprocessing

for x in train.columns:
    if train[x].dtype == 'object':
        lbl = preprocessing.LabelEncoder()
        lbl.fit(list(train[x].values))
        train[x] = lbl.transform(list(train[x].values))

In [271]:
# Check the changes
train.head()

age  workclass  fnlwgt  education  education.num  marital.status  \
0   39          6   77516          9             13               4   
1   50          5   83311          9             13               2   
2   38          3  215646         11              9               0   
3   53          3  234721          1              7               2   
4   28          3  338409          9             13               2   

   occupation  relationship  race  sex  capital.gain  capital.loss  \
0           0             1     4    1          2174             0   
1           3             0     4    1             0             0   
2           5             1     4    1             0             0   
3           5             0     2    1             0             0   
4           9             5     2    0             0             0   

   hours.per.week  native.country  target  
0              40              38       0  
1              13              38       0  
2              40              38       0  
3              40              38       0  
4              40               4       0

As we can see, all the variables have been converted to numeric, including the target variable.

In [272]:
# <50k = 0 and >50k = 1
train.target.value_counts()

target
0    24720
1     7841
Name: count, dtype: int64

## Building a Random Forest Model

In [273]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

y = train['target']
del train['target']

X = train
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1, stratify=y)

# train the RF classifier
clf = RandomForestClassifier(n_estimators=500, max_depth=6)
clf.fit(X_train, y_train)

clf.predict(X_test)

array([0, 1, 0, ..., 0, 0, 0])

In [274]:
# make prediction and check model's accuracy
prediction = clf.predict(X_test)
acc = accuracy_score(np.array(y_test), prediction)
print('The accuracy of Random Forest is {}', format(acc))

The accuracy of Random Forest is {} 0.8525949431876344


Hurrah! Our learning algorithm gave 85% accuracy.